state/upgrades.go

```diff
@@ -2246,3 +2246,72 @@ func AddControllerNodeDocs(pool *StatePool) error {
 	}
 	return nil
 }
+
+// AddSpaceIdToSpaceDocs ensures that every space document includes a
+// a sequentially generated ID.
+// It also adds a doc for the default space (ID=0).
+func AddSpaceIdToSpaceDocs(pool *StatePool) (err error) {
+	return errors.Trace(runForAllModelStates(pool, func(st *State) error {
+		// Make sure we do not add a default space to the controller model.
+		if st.isController() {
+			return nil
+		}
+
+		col, closer := st.db().GetCollection(spacesC)
+		defer closer()
+
+		type oldSpaceDoc struct {
+			SpaceId    string `bson:"spaceid"`
+			Life       Life   `bson:"life"`
+			Name       string `bson:"name"`
+			IsPublic   bool   `bson:"is-public"`
+			ProviderId string `bson:"providerid,omitempty"`
+		}
+
+		var docs []oldSpaceDoc
+		err := col.Find(nil).All(&docs)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		var ops []txn.Op
+		for _, oldDoc := range docs {
+			// A doc with a space ID has already been upgraded.
+			if oldDoc.SpaceId != "" {
+				continue
```

Comment:
I notice you aren't tracking names that have upgraded SpaceIds. Does that mean if this gets partially done and then restarted (say the docs got created but somehow the old ones not removed), you would end up with 2 possible SpaceIds for the same SpaceName ?
I realize if all the ops are run in a single TXN then it shouldn't ever get partially applied.
Just a defensive programming stance. Maybe it isn't worth worrying about (you also have to watch out for seeing the document with SpaceId *after* you've already tried to create one.)